<a href="https://colab.research.google.com/github/tjsgertsen/Chainer_repro/blob/master/Chainer_repro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Reproduction by using Chainer

To reproduce the results of the paper 'Between-class Learning for Image Classification' the code is made available by the authors. However, this code is written in python 2 with a very early version of the Chainer library. Since we had to work on Google Colabs in order to make use of GPU computing this old versions did not work anymore. Therefore, we decided to approach the reproduction in two ways: 
1. by building our own version in Pytorch (other google colab)
2. Update the original code from the authors by making it compatible with python 3 and the newest version of Chainer. To keep the structure of the code intact it was modified offline whereafter is loaded into this Google Colab. Furthermore, an effort is made to change as little as possible since the goal is to reproduce. The (offline) code can be found in the following Github repository: [https://github.com/tjsgertsen/Chainer_repro.git](https://)

First, your google drive is mounted. Change `root_path` to your folder where the (offline) code is stored in google drive. You still have to download the CIFAR-10 into a folder named "datasets" within the same folder as the code.



In [21]:
from google.colab import drive
import os
import chainer

# mount drive
drive.mount('/content/gdrive', force_remount=True) 

# change working directory
root_path='/content/gdrive/My Drive/Cloud Sync/bc_learning_image-master_python3' 
os.chdir(root_path)
check = os.getcwd()
print(f'Check: the current working directory is {check}')

!python --version
print(f'Chainer {chainer.__version__}')


Mounted at /content/gdrive
Check: the current working directory is /content/gdrive/My Drive/Cloud Sync/bc_learning_image-master_python3
Python 3.6.9
Chainer 6.5.0


# Run Script

In order to initialize the code the following python command is triggered. A few examples for variatonal settings are commented. Please construct your own problem that you want to reproduce or run:

In [0]:
print(f'Check chainer version: {chainer.__version__}')
%matplotlib inline

# command to run with cifar-10 dataset and standard method
# !python main.py --dataset cifar10 --netType convnet --nTrials 3 --data /content/gdrive/My\ Drive/Cloud\ Sync/bc_learning_image-master_python3/datasets/cifar-10-batches-py

# command to run with cifar-10 dataset and BC(+) method
!python main.py --dataset cifar10 --netType convnet --nTrials 2 --data /content/gdrive/My\ Drive/Cloud\ Sync/bc_learning_image-master_python3/datasets/cifar-10-batches-py --BC --plus \
--save /content/gdrive/My\ Drive/Cloud\ Sync/bc_learning_image-master_python3/results

# command to run with cifar-100 dataset
#!python main.py --dataset cifar100 --netType convnet --data /content/gdrive/My Drive/Cloud Sync/bc_learning_image-master_python3/datasets/cifar-100-python --BC --plus

Check chainer version: 6.5.0
+------------------------------+
| CIFAR classification
+------------------------------+
| dataset  : cifar10
| netType  : convnet
| learning : BC+
| nEpochs  : 250
| LRInit   : 0.1
| schedule : [0.4, 0.6, 0.8]
| warmup   : 0
| batchSize: 128
+------------------------------+
+-- Trial 1 --+
| Epoch: 1/250 | Train: LR 0.1  Loss 1.574  top1 73.58 | Val: top1 53.41
| Epoch: 2/250 | Train: LR 0.1  Loss 1.316  top1 54.02 | Val: top1 40.44
| Epoch: 3/250 | Train: LR 0.1  Loss 1.150  top1 43.42 | Val: top1 30.04
| Epoch: 4/250 | Train: LR 0.1  Loss 1.056  top1 37.71 | Val: top1 38.87
| Epoch: 5/250 | Train: LR 0.1  Loss 0.993  top1 34.35 | Val: top1 27.74
| Epoch: 6/250 | Train: LR 0.1  Loss 0.959  top1 32.65 | Val: top1 25.48
| Epoch: 7/250 | Train: LR 0.1  Loss 0.930  top1 31.02 | Val: top1 25.05
| Epoch: 8/250 | Train: LR 0.1  Loss 0.908  top1 29.74 | Val: top1 21.30
| Epoch: 9/250 | Train: LR 0.1  Loss 0.889  top1 28.73 | Val: top1 19.35
| Epoch: 10/250 | Trai

# Results

Since the runtime is restricted by Google Colabs we tried to reproduce table 2 of the paper for all learning methods with the CIFAR-10 dataset and an 11-layer CNN. The standard deviation is constructed by 5 trials and 250 epoch per trial, of each learning method. The results that are found are as following:

Learning method | Paper results | Reproduction
--- | --- | ---
Standard | $6.07 \pm 0.04$ | $6.11 \pm 0.15$
BC | $5.40 \pm 0.07$ | $5.39 \pm 0.07$
BC+ | $5.22 \pm 0.04$ | $5.26 \pm 0.16$